### ISSUE TITLE
Android: Mdbx Error(11)

### ISSUE THREAD

**Issue by robotoss** [isar/isar#570]
**Steps to Reproduce**
On real Android device.
1. Run release application.
2. Await when Isar is open.
3. Exit from app by tap Android back button.
4. Repeat steps 3 through 1 several times in a row.

*Not sure exactly, but it is possible that the application is quit while writingTxn is running
** No error if we exit by exit(0); But it gives us other problems with Oppo and Samsung phones that is not a solution.

**Version**
 - Platform: Android
 - Flutter version: 3.0.5
 - Isar version: 3.0.0-dev.8
 - Isar Generator version: 3.0.0-dev.8

 

https://user-images.githubusercontent.com/34773197/185613036-d22aac3d-7b42-4a35-9a28-ac3dceb15dd6.mp4

**Comment by leisim** [comment]
Do you have an error message, stack trace or steps to reproduce? Without that it is really difficult to help

**Comment by robotoss** [comment]
@leisim 
I will try to make a test open-source project today which would reproduce the error

**Comment by leisim** [comment]
@robotoss That would be incredibly helpful! Thank you!

**Comment by robotoss** [comment]
@leisim 
Good day.
Sorry for the delay.

https://github.com/robotoss/rick_and_morty_isar
We have a very confusing data structure that the api service gives us, so we had to add a delay (it's the only one where the problems start)
https://github.com/robotoss/rick_and_morty_isar/commit/cac77f90562c901bf555a3db3e0ae6a12cf5196e

To play the bug you need to run on Android in release, and press the native back button when the data is loaded.
After several attempts the bug appears.
I have not noticed this on iOS


https://user-images.githubusercontent.com/34773197/185781585-0ee7426a-8d77-4166-8664-89c24895974e.mp4

**Comment by leisim** [comment]
@robotoss Thanks a lot! This looks to me like the previous process is still blocking the database file so the new Isar instance cannot access it.

**Comment by leisim** [comment]
@robotoss do you only observe this behavior only after killing and restarting the app (with the back button) or also while the app is running?

**Comment by robotoss** [comment]
During the restart via the back button and a couple of times it appeared during the hot-restart

**Comment by leisim** [comment]
Okay then the most likely explanation is that Android does not immediately release the file lock or the mmaped memory. I'm not sure if/how we can fix that

**Comment by robotoss** [comment]
Good day, I add while to isar init.
https://github.com/robotoss/rick_and_morty_isar/commit/af47da1a7f2a23e8b5c52a2dfac249bb8aa4b226
But it is doesn't help.

**Comment by Chappie74** [comment]
Hey I tested 3.0.0-dev.12 on x86  API 29 and it works just fine. 
Same thing on api 30 throws this error for me. Maybe that can help you investigate the problem.

**Comment by leisim** [comment]
That's super helpful thanks @Chappie74. My issue right now is that M1 macs cannot run x86 emulators 😞

**Comment by Macacoazul01** [comment]
> Okay then the most likely explanation is that Android does not immediately release the file lock or the mmaped memory. I'm not sure if/how we can fix that.
> 
> I wonder how common this happens in real life. I would suggest to wrap the Isar.open() in try/catch and just retry after 500ms

@leisim i'm having the same error here after trying to run a hot restart during an error that happened after not waiting properly for `isar.writeTxn` to end (the first time it didn't appear, but after repeating the process some times it did).

So maybe it's related to some memory lock as you said.

Closing the debugger and starting it again solved my problem here.

**Comment by AdilMango** [comment]
Hello @simc , i am facing this issue .
version i am using isar: ^3.0.5

**Comment by dayron9110** [comment]
Same error

**Comment by dayron9110** [comment]
I have this problem using a singleton from the db with get_it

**Comment by gausoft** [comment]
> while



> Same error, I am using isar 3.0.5

Me too

**Comment by brendan-dexter** [comment]
This is still a major problem! I wonder why it was marked as closed.

**Comment by erthink** [comment]
I am the main developer and maintainer of `libmdbx` (the key-value storage engine).

It's amazing to me that in 3 years (!) no one has asked me at least a question about this. For more than 5 years, there has been a publicly accessible telegram group, the link to which is on [libmdbx.dqdkfa.ru](https://libmdbx.dqdkfa.ru/) (the project's home site).

At the moment, I am not aware of any error or flaw in libmdbx that could lead to the described issue.

Moreover, the described situation is not a problem. The error code `11` corresponds to the `EAGAIN` system error code, which is routinely returned by the system kernel in a number of situations when some resource is occupied or insufficient (for example, not enough free RAM or a file region is locked).

### From a practical point of view, libmdbx usage scenarios on Android can potentially have the following situations:
1. There is not enough free RAM and/or PTE (page table entries) to map the database into memory.
2. The process previously working with the database has not yet managed to close the database and/or terminate.
In particular, it should be borne in mind that closing the database may require writing some data to the storage medium, which may be a lot in certain operating scenarios and/or non-optimal/non-optimal use of the database.
Such writing may overlap with other activity/load, or require performing maintenance operations (garbage collection, recovery, transfer/copy) at the file system and/or flash memory controller level.
In the rare worst cases, the delay peaks/spikes can be seconds, and when the flash drive is tired, they can take up to tens of seconds.
3. The system resources used on the previous startup are not yet available for use.
Some stages of virtual memory management in the OS kernel are performed on a delayed basis, in lower priority threads, or no more often than a certain pace (once per second).
On Android/iOS, there are more system mechanisms involved, including managing/clearing various kind of system caches, etc.
So, the system kernel/services can return `EAGAIN` when there are no obvious obvious reasons for this.
4. Antiviruses or other similar services can analyze a files mapped in RAM (i.e. a DB in libmdbx case), both before opening and after closing ones. This can happen both synchronously and asynchronously in the background. This can also lead to a temporary lack of RAM or an explicit return of the error EAGAIN when opening, if the last such analysis has not finished yet.
5. The larger the database size, both current and maximum (set in the database geometry), the more RAM and PTA will be required and (therefore) the greater the probability of manifestation of discussed issue.

### As a workaround, I can recommend the following:
1. Captain obvious: The capabilities of libmdbx should be used rationally, especially the modes of operation of the database and the API for explicitly flushing data to storage media.
2. During the termination of an application, close the MDBX database explicitly and as early as possible.
3. In case the `EAGAIN` error when opening a database, try again a reasonable number of times with a short delay (for example, 10 with a pause from 0.01 to 0.1 seconds).

If someone wants to investigate the cause(s) more deeply and/or find better workaround, then I will give the necessary advices/suggestions (mainly get the debugging version from me and enable logging). Please ask me in libmdbx telegram group.

Nonetheless, please use the latest libmdbx release on the origin above, either on [github-mirror](https://github.com/erthink/libmdbx). Ask the maintainers and/or developers of Isar to update libmdbx used and/or contact the maintainers/developers of the bindings used to update the used library, etc.
